tensorflow/core/ops/batchnorm_training_op.cc

```diff
@@ -0,0 +1,80 @@
+/* Copyright 2015 Google Inc. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/core/framework/op.h"
+
+namespace tensorflow {
+
+REGISTER_OP("BatchNormalizeTraining")
+    .Input("input: T")
+    .Input("scale: T")
+    .Input("bias: T")
+    .Input("running_mean: Ref(T)")
+    .Input("running_inv_var: Ref(T)")
+    .Output("out: T")
+    .Output("save_mean: T")
+    .Output("save_inv_var: T")
+    .Attr("T: {float}")
+    .Attr("epsilon: float")
+    .Attr("exponential_average_factor: float")
```

Comment:
When shifting to T or double, somewhere in TF's code generation fails. Does this make sense? An alternative could be to make epsilon a tensor input and not an attr but that seems a little extreme.

```
  /bin/bash -c 'source external/bazel_tools/tools/genrule/genrule-setup.sh; bazel-out/host/bin/tensorflow/python/gen_batchnorm_training_op_py_wrappers_cc BatchNormTrainingGrad 0 > bazel-out/local_linux-opt/genfiles/tensorflow/python/gen_batchnorm_training_op.py').
F tensorflow/core/framework/op.cc:132] Check failed: ::tensorflow::Status::OK() == (RegisterAlreadyLocked(op_def)) (OK vs. Invalid argument: Unrecognized type '' in attr 'epsilon'; in OpDef: name: "BatchNormTraining" input_arg { name: "input" description: "A 4D input Tensor." type_attr: "T" } input_arg { name: "scale" description: "A 1D Tensor with size equal to the number of channels.\nThis is the learned scale value multiplied post normalization.\nAlso known as gamma." type_attr: "T" } input_arg { name: "bias" description: "A 1D Tensor with size equal to the number of channels.\nThis is the learned bias value added post normalization.\nAlso known as beta." type_attr: "T" } output_arg { name: "out" description: "A 4D output Tensor. The input after applying batch normalization." type_attr: "T" } output_arg { name: "save_mean" description: "A 1D Tensor. Computed means. To be used in the backward pass." type_attr: "T" } output_arg { name: "save_inv_var" description: "A 1D Tensor. Computed inverse variance. To be used in the backward pass." type_attr: "T" } attr { name: "T" type: "type" allowed_values { list { type: DT_FLOAT type: DT_DOUBLE } } } attr { name: "epsilon" description: "float" } attr { name: "data_format" type: "string" } summary: "Perform batch norm using batch statistics.")Attempting to register: Op<name=BatchNormTraining; signature=input:T, scale:T, bias:T -> out:T, save_mean:T, save_inv_var:T; attr=T:type,allowed=[DT_FLOAT, DT_DOUBLE]; attr=epsilon:; attr=data_format:string>
```